src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newCopy = { ...state };
+  const newArray = [];
+
+  const ADD_PROP = 'addProperties';
+  const REMOVE_PROP = 'removeProperties';
+  const CLEAR_PROP = 'clear';
+
+  for (const action of actions) {
+    let cloneState = { ...newCopy };
+
+    switch (action.type) {
+      case ADD_PROP:
+        Object.assign(cloneState, action.extraData);
+
+        break;
+
+      case REMOVE_PROP: {
+        for (const key of action.keysToRemove) {
+          delete cloneState[key];
+        }
+
+        break;
+      }
+
+      case CLEAR_PROP: {
+        cloneState = {};
+
+        break;
+      }
+    }
+    newArray.push(cloneState);
+    newCopy = cloneState;
+  }
```

Comment:
you dont need to create new copy  let cloneState = { ...newCopy };
```suggestion
  for (const action of actions) {
    switch (action.type) {
      case ADD_PROP:
        Object.assign(newCopy, action.extraData);

        break;

      case REMOVE_PROP: {
        for (const key of action.keysToRemove) {
          delete newCopy[key];
        }

        break;
      }

      case CLEAR_PROP: {
        newCopy = {};

        break;
      }
    }
    newArray.push({...newCopy });
  }
```